File: articles/personality-wellbeing-and-mental-health/r/analyze_personality_wellbeing.R
# Personality, Wellbeing, and Mental Health
# R workflow for trait, distress, flourishing, and support analysis

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_personality_wellbeing_mental_health.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

required <- c(
  "participant_id",
  "age_band",
  "life_context",
  "extraversion",
  "agreeableness",
  "conscientiousness",
  "neuroticism",
  "openness",
  "coping_effectiveness",
  "perceived_support",
  "stress_burden",
  "positive_affect",
  "negative_affect",
  "life_satisfaction",
  "meaning_purpose",
  "wellbeing_score",
  "distress_score",
  "flourishing_score",
  "social_functioning",
  "treatment_access",
  "sleep_quality"
)

missing <- setdiff(required, names(data))
if (length(missing) > 0) {
  stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
}

data$regulation_support_index <- (
  data$conscientiousness +
  data$coping_effectiveness +
  data$perceived_support +
  data$sleep_quality -
  data$stress_burden
) / 4

data$flourishing_capacity_index <- (
  data$positive_affect +
  data$meaning_purpose +
  data$life_satisfaction +
  data$perceived_support -
  data$negative_affect
) / 4

data$distress_vulnerability_index <- (
  data$neuroticism +
  data$stress_burden +
  data$negative_affect -
  data$coping_effectiveness -
  data$perceived_support
)

data$distress_level <- ifelse(
  data$distress_score > median(data$distress_score, na.rm = TRUE),
  "higher_distress",
  "lower_distress"
)

data$flourishing_level <- ifelse(
  data$flourishing_score > median(data$flourishing_score, na.rm = TRUE),
  "higher_flourishing",
  "lower_flourishing"
)

data$mental_health_profile <- paste(
  data$distress_level,
  data$flourishing_level,
  sep = "_"
)

context_summary <- aggregate(
  data[, c("wellbeing_score", "distress_score", "flourishing_score", "perceived_support", "stress_burden", "treatment_access")],
  by = list(life_context = data$life_context),
  FUN = mean
)

context_n <- aggregate(
  data$participant_id,
  by = list(life_context = data$life_context),
  FUN = length
)
names(context_n)[2] <- "n"
context_summary <- merge(context_n, context_summary, by = "life_context")

write.csv(context_summary, file.path(outputs, "r_life_context_summary.csv"), row.names = FALSE)

profile_summary <- aggregate(
  data[, c("wellbeing_score", "distress_score", "flourishing_score", "neuroticism", "extraversion", "conscientiousness", "perceived_support")],
  by = list(mental_health_profile = data$mental_health_profile),
  FUN = mean
)

profile_n <- aggregate(
  data$participant_id,
  by = list(mental_health_profile = data$mental_health_profile),
  FUN = length
)
names(profile_n)[2] <- "n"
profile_summary <- merge(profile_n, profile_summary, by = "mental_health_profile")

write.csv(profile_summary, file.path(outputs, "r_two_continua_profile_summary.csv"), row.names = FALSE)

corr_cols <- c(
  "extraversion",
  "agreeableness",
  "conscientiousness",
  "neuroticism",
  "openness",
  "coping_effectiveness",
  "perceived_support",
  "stress_burden",
  "positive_affect",
  "negative_affect",
  "life_satisfaction",
  "meaning_purpose",
  "wellbeing_score",
  "distress_score",
  "flourishing_score",
  "social_functioning"
)

cor_matrix <- cor(data[, corr_cols], use = "pairwise.complete.obs")
write.csv(cor_matrix, file.path(outputs, "r_trait_wellbeing_correlations.csv"))

model_wb <- lm(
  wellbeing_score ~ extraversion + agreeableness +
    conscientiousness + neuroticism + openness +
    perceived_support + coping_effectiveness + sleep_quality,
  data = data
)

model_distress <- lm(
  distress_score ~ extraversion + agreeableness +
    conscientiousness + neuroticism + openness +
    stress_burden + perceived_support + sleep_quality,
  data = data
)

model_flourishing <- lm(
  flourishing_score ~ extraversion + agreeableness +
    conscientiousness + neuroticism + openness +
    meaning_purpose + perceived_support + coping_effectiveness,
  data = data
)

model_social <- lm(
  social_functioning ~ extraversion + agreeableness +
    conscientiousness + perceived_support + distress_score + sleep_quality,
  data = data
)

capture.output(summary(model_wb), file = file.path(outputs, "r_wellbeing_model_summary.txt"))
capture.output(summary(model_distress), file = file.path(outputs, "r_distress_model_summary.txt"))
capture.output(summary(model_flourishing), file = file.path(outputs, "r_flourishing_model_summary.txt"))
capture.output(summary(model_social), file = file.path(outputs, "r_social_functioning_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_personality_wellbeing.csv"), row.names = FALSE)

cat("Wrote R outputs to:", outputs, "\n")

File: articles/personality-wellbeing-and-mental-health/python/analyze_personality_wellbeing.py
#!/usr/bin/env python3
"""Analyze synthetic personality, wellbeing, and mental-health data.

This script supports the article "Personality, Wellbeing, and Mental Health."
It demonstrates age-band summaries, life-context summaries, trait-wellbeing
correlations, two-continua mental-health profiles, and regression-style models
using synthetic data only.
"""

from __future__ import annotations

from pathlib import Path
import json
import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "synthetic_personality_wellbeing_mental_health.csv"
OUTPUTS = ROOT / "outputs"

TRAITS = [
    "extraversion",
    "agreeableness",
    "conscientiousness",
    "neuroticism",
    "openness",
]

PATHWAYS = [
    "coping_effectiveness",
    "perceived_support",
    "stress_burden",
    "treatment_access",
    "sleep_quality",
]

OUTCOMES = [
    "wellbeing_score",
    "distress_score",
    "flourishing_score",
    "social_functioning",
]


def fit_ols(df: pd.DataFrame, y_col: str, x_cols: list[str]) -> dict[str, float]:
    """Fit a small OLS model with numpy and return coefficients."""
    X = df[x_cols].to_numpy(dtype=float)
    X = np.column_stack([np.ones(len(X)), X])
    y = df[y_col].to_numpy(dtype=float)
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)
    names = ["intercept", *x_cols]
    return {name: float(value) for name, value in zip(names, beta)}


def main() -> None:
    OUTPUTS.mkdir(parents=True, exist_ok=True)

    df = pd.read_csv(DATA_PATH)

    required = {
        "participant_id",
        "age_band",
        "life_context",
        *TRAITS,
        *PATHWAYS,
        "positive_affect",
        "negative_affect",
        "life_satisfaction",
        "meaning_purpose",
        *OUTCOMES,
    }
    missing = required.difference(df.columns)
    if missing:
        raise ValueError(f"Missing required columns: {sorted(missing)}")

    df["regulation_support_index"] = (
        df["conscientiousness"]
        + df["coping_effectiveness"]
        + df["perceived_support"]
        + df["sleep_quality"]
        - df["stress_burden"]
    ) / 4.0

    df["flourishing_capacity_index"] = (
        df["positive_affect"]
        + df["meaning_purpose"]
        + df["life_satisfaction"]
        + df["perceived_support"]
        - df["negative_affect"]
    ) / 4.0

    df["distress_vulnerability_index"] = (
        df["neuroticism"]
        + df["stress_burden"]
        + df["negative_affect"]
        - df["coping_effectiveness"]
        - df["perceived_support"]
    )

    df["distress_level"] = np.where(
        df["distress_score"] > df["distress_score"].median(),
        "higher_distress",
        "lower_distress",
    )

    df["flourishing_level"] = np.where(
        df["flourishing_score"] > df["flourishing_score"].median(),
        "higher_flourishing",
        "lower_flourishing",
    )

    df["mental_health_profile"] = (
        df["distress_level"] + "_" + df["flourishing_level"]
    )

    context_summary = (
        df.groupby("life_context")
        .agg(
            n=("participant_id", "count"),
            wellbeing_mean=("wellbeing_score", "mean"),
            distress_mean=("distress_score", "mean"),
            flourishing_mean=("flourishing_score", "mean"),
            support_mean=("perceived_support", "mean"),
            stress_burden_mean=("stress_burden", "mean"),
            treatment_access_mean=("treatment_access", "mean"),
        )
        .reset_index()
    )

    age_summary = (
        df.groupby("age_band")
        .agg(
            n=("participant_id", "count"),
            positive_affect_mean=("positive_affect", "mean"),
            negative_affect_mean=("negative_affect", "mean"),
            life_satisfaction_mean=("life_satisfaction", "mean"),
            meaning_purpose_mean=("meaning_purpose", "mean"),
            social_functioning_mean=("social_functioning", "mean"),
        )
        .reset_index()
    )

    profile_summary = (
        df.groupby("mental_health_profile")
        .agg(
            n=("participant_id", "count"),
            wellbeing_mean=("wellbeing_score", "mean"),
            distress_mean=("distress_score", "mean"),
            flourishing_mean=("flourishing_score", "mean"),
            neuroticism_mean=("neuroticism", "mean"),
            extraversion_mean=("extraversion", "mean"),
            conscientiousness_mean=("conscientiousness", "mean"),
            perceived_support_mean=("perceived_support", "mean"),
        )
        .reset_index()
    )

    context_summary.to_csv(OUTPUTS / "python_life_context_summary.csv", index=False)
    age_summary.to_csv(OUTPUTS / "python_age_band_summary.csv", index=False)
    profile_summary.to_csv(OUTPUTS / "python_two_continua_profile_summary.csv", index=False)

    corr_cols = TRAITS + PATHWAYS + [
        "positive_affect",
        "negative_affect",
        "life_satisfaction",
        "meaning_purpose",
        "regulation_support_index",
        "flourishing_capacity_index",
        "distress_vulnerability_index",
    ] + OUTCOMES

    df[corr_cols].corr().to_csv(OUTPUTS / "python_trait_wellbeing_correlations.csv")

    wellbeing_inputs = TRAITS + ["perceived_support", "coping_effectiveness", "sleep_quality"]
    distress_inputs = TRAITS + ["stress_burden", "perceived_support", "sleep_quality"]
    flourishing_inputs = TRAITS + ["meaning_purpose", "perceived_support", "coping_effectiveness"]
    functioning_inputs = ["extraversion", "agreeableness", "conscientiousness", "perceived_support", "distress_score", "sleep_quality"]

    models = {
        "wellbeing_score": fit_ols(df, "wellbeing_score", wellbeing_inputs),
        "distress_score": fit_ols(df, "distress_score", distress_inputs),
        "flourishing_score": fit_ols(df, "flourishing_score", flourishing_inputs),
        "social_functioning": fit_ols(df, "social_functioning", functioning_inputs),
        "note": "Synthetic demonstration only; not a clinical, diagnostic, or mental-health scoring model.",
    }

    with open(OUTPUTS / "python_personality_wellbeing_models.json", "w", encoding="utf-8") as f:
        json.dump(models, f, indent=2)

    df.to_csv(OUTPUTS / "python_scored_personality_wellbeing.csv", index=False)

    print("Wrote Python outputs to:", OUTPUTS)


if __name__ == "__main__":
    main()
